### PR TITLE
interchange: refactor Avro Encoder

### PR DESCRIPTION
**Pull request by JLDLaughlin**
Creates a new generic Encoder Trait that will be implemented by
multiple encoder types (Avro, JSON, etc.).

### CODE REVIEW

**Comment by JLDLaughlin** [comment]
@quodlibetor -- does this seem reasonable to you? The goal is to add a `JsonEncoder` to [enable JSON formatted Kafka sinks](https://github.com/MaterializeInc/materialize/issues/1540).

File: src/dataflow/src/sink/kafka.rs [hunk 1]
```diff
@@ -694,15 +697,14 @@ fn avro_encode_stream<G>(
...
-    key_schema_id: Option<i32>,
-    value_schema_id: i32,
+    encoder: impl Encode + 'static,
     fuel: usize,
     name_prefix: String,
 ) -> Stream<G, ((Option<Vec<u8>>, Option<Vec<u8>>), Timestamp, Diff)>
 where
     G: Scope<Timestamp = Timestamp>,
 {
+    // todo change name
```
Comment (aljoscha) [inline src/dataflow/src/sink/kafka.rs:707]: I'm guessing you want to put something like `get_format_name()` in the `Encode` trait? I'm still not experienced enough to have opinions about good trait design.

File: src/dataflow/src/sink/avro_ocf.rs [hunk 2]
```diff
@@ -36,7 +36,7 @@ pub fn avro_ocf<G>(
         v
     });
     let (schema, columns) = {
-        let encoder = Encoder::new(None, desc, false);
+        let encoder = AvroEncoder::new(None, desc, false, None, None);
```
Comment (aljoscha) [inline src/dataflow/src/sink/avro_ocf.rs:39]: This made me remember that the old `Encoder` is weird and has this double-use we see here. Where the schema IDs are `None`, which feels iffy.

File: src/sql/src/plan/statement/ddl.rs [hunk 3]
```diff
@@ -1312,12 +1312,14 @@ fn kafka_sink_builder(
         Vec::new()
     };
 
-    let encoder = Encoder::new(
+    let encoder = AvroEncoder::new(
```
Comment (aljoscha) [inline src/sql/src/plan/statement/ddl.rs:1315]: Same as above, it feels weird that the `Encoder` is also used for this.

File: src/interchange/src/avro/encode.rs [hunk 4]
```diff
@@ -213,6 +221,16 @@ impl Encoder {
     }
 }
 
+impl Encode for AvroEncoder {
+    fn encode_key_unchecked(&self, row: Row) -> Vec<u8> {
+        self.encode_key_unchecked(self.key_schema_id.unwrap(), row)
```
Comment (aljoscha) [inline src/interchange/src/avro/encode.rs:226]: I don't like the `unwraps`, but getting rid of them seems even more complicated.

**Review by aljoscha — APPROVED** [review]
I like this! I had some inline comments but you can interpret all of them as nits. Though eventually we will have to tighten things a bit around the encoders, I feel.

File: src/dataflow/src/sink/avro_ocf.rs (hunk 2 above)
Comment (JLDLaughlin) [reply src/dataflow/src/sink/avro_ocf.rs:39]: Agreed! This bothered me too... I think I'll separate them out!

File: src/interchange/src/avro/encode.rs (hunk 4 above)
Comment (JLDLaughlin) [reply src/interchange/src/avro/encode.rs:226]: Agreed! I'm going to play with making a second Encoder, which might help.

**Comment by JLDLaughlin** [comment]
@aljoscha -- I made a `AvroSchemaGenerator` to break up the responsibilities of the `AvroEncoder`. It addresses some of our concerns, but not all of them!

File: src/interchange/src/avro/encode.rs [hunk 5]
```diff
@@ -36,7 +38,7 @@ lazy_static! {
         "name": "envelope",
         "fields": [
             {
-                "name": "id",
+            "name": "id",
```
Comment (aljoscha) [inline src/interchange/src/avro/encode.rs:41]: seems unintended

File: src/interchange/src/avro/encode.rs [hunk 6]
```diff
@@ -199,20 +193,71 @@ impl Encoder {
...
+/// Manages encoding of Avro-encoded bytes.
+pub struct AvroEncoder {
+    schema_generator: AvroSchemaGenerator,
+    key_schema_id: Option<i32>,
+    value_schema_id: i32,
+}
+
+impl fmt::Debug for AvroEncoder {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("Encoder")
```
Comment (aljoscha) [inline src/interchange/src/avro/encode.rs:215]: I don't know if the convention is to do `"AvroEncoder"` or `"Encoder"`, here.

**Review by aljoscha — APPROVED** [review]
Thanks, that does clean it up a bit!

**Pull request merged**